Morph rework and check spell characteristics

ssk97 opened the pull request:
Morph should be a separate SpellAbility rather than an AlternativeSourceCostsImpl. Changing this allows me to support the interaction with [[Liberator, Urza's Battlethopter]] and remove a lot of the code special casing [[Zoetic Cavern]].

Also, changed tests which "Cast X" to require the exact name, instead of just checking the start. Previously it allowed casting with alternative modes without actually naming them, which doesn't seem like good behavior.

Fixes #10168 
Fixes #11162 
Fixes #7160
Fixes #7121 
Fixes #4294
Fixes #7138

~Needs more testing and is currently missing some more of the abilities that should be checking spell characteristics rather than the card's, but the basic idea is complete.~ Completed

This also allows me to convert AlternativeSourceCostsImpl to use CostTags once that change be merged, since Morph doesn't fit well with it as-is.

jeffwadsworth commented:
Morph is a special action that doesn't use the stack.  Is this covered in your code?

xenohedron commented:
I like the direction

Qumojo commented:
Not sure if I should add it here or create a new thread but here it goes. When I play a morph card (I am a newbie, learning slowly with my friend) then I can't check the card and some times forget what they all do/are! Could we get an option when moving over the card with the mouse the original side will show up instead of that generic morph? Thanks

ssk97 commented:
Manual testing suggests that in gameplay terms my changes function correctly, though I didn't test every card affected by the changes to checking characteristics. However, there are some remaining UI issues.

The creature seems to keep the cast ability while on the battlefield
![image](https://github.com/magefree/mage/assets/8648826/78395332-f981-49df-9c8a-3fe8604472ea)
I'll keep this PR in draft mode until I fix this (unless instructed otherwise)

Also, on the "choose how to play card" popup the Morph text is a little long:
![image](https://github.com/magefree/mage/assets/8648826/13bb0a6f-726e-417d-9c4f-18b047732145)
with the entire reminder text being there if you scroll. Not sure if this is a problem or not.

> Could we get an option when moving over the card with the mouse the original side will show up instead of that generic morph?

That's definitely something that should go in a separate issue/pull request. It sounds like a good idea, but would require a good bit of UI work.

xenohedron commented:
should fix #7160 ?

ssk97 commented:
Fixed the cast ability (it incorrectly had setWorksFaceDown set) and removed SpellAbilityType.FACE_DOWN_CREATURE since it was only used for Zoetic Cavern and is thus no longer needed.

I believe this PR is now ready for review.

xenohedron commented on Mage/src/main/java/mage/abilities/SpellAbility.java:
```diff
@@ -306,7 +306,7 @@ public Card getCharacteristics(Game game) {
 
     public static SpellAbility getSpellAbilityFromEvent(GameEvent event, Game game) {
-        if (event.getType() != GameEvent.EventType.CAST_SPELL) {
+        if (event.getType() != GameEvent.EventType.CAST_SPELL && event.getType() != GameEvent.EventType.CAST_SPELL_LATE) {
```
can you explain this change?

ssk97 replied:
Getting the name of a spell for comparison happens in CAST_SPELL_LATE. I'm not sure what this function used to be used for (it currently had no usage) but since I needed basically this function and it was already in the code, I used it with the small change.

xenohedron commented on Mage/src/main/java/mage/abilities/SpellAbility.java:
```diff
@@ -306,7 +306,7 @@ public Card getCharacteristics(Game game) {
     }
 
     public static SpellAbility getSpellAbilityFromEvent(GameEvent event, Game game) {
```
it seems this method is used in a bunch of places for when comparing names as morphs have no name. Can you add some sort of explanatory comment on its usage?

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/continuous/CastAsThoughItHadFlashAllEffect.java:
```diff
@@ -50,25 +49,19 @@ public CastAsThoughItHadFlashAllEffect copy() {
         return false;
     }
+    @Override
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
```
can you explain why this method is being split into two with different signatures?

ssk97 replied:
I need access to the affectedAbility argument. The parent class has the full applies method calling the shorter one by default. However, instead of returning false, I should be throwing an IllegalArgumentException.

xenohedron commented on Mage.Sets/src/mage/cards/c/ChainerNightmareAdept.java:
```diff
@@ -92,18 +93,21 @@ public void init(Ability source, Game game) {
     @Override
-    public boolean applies(UUID sourceId, Ability source, UUID affectedControllerId, Game game) {
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        throw new IllegalArgumentException("ERROR, can't call applies method on empty affectedAbility");
```
style: prefer "Wrong code usage:" to "ERROR," (multiple places)

xenohedron commented:
Two more issues to check that I think might be resolved by this rework
* #9510
* #7138

xenohedron reviewed (commented):
Thanks for separating out into logical commits, it really helped streamline the review. Basically, all the changes look good to me. There are a couple minor style things that should be easy to address. The main thing I'd like to see is some sort of documentation to guide future developers into correctly using the spell characteristics check when applicable.

ssk97 commented:
> Two more issues to check that I think might be resolved by this rework

Second issue is definitely fixed by making morph involved in the characteristics checking.

First issue seems to have two parts. The first is fixed while the second is not. Fixing that I think involves moving the "can't cast" effects from CAST_SPELL_LATE to CAST_SPELL events, but there's probably other bits involved that I haven't investigated.

JayDi85 commented:
> However, there are some remaining UI issues
> The creature seems to keep the cast ability while on the battlefield

If you allows to activate it then it's not GUI issue, it's a critical game engine issue and must be fixed.

> Also, on the "choose how to play card" popup the Morph text is a little long with the entire reminder text being there if you scroll. Not sure if this is a problem or not

It's ok

ssk97 commented:
> If you allows to activate it then it's not GUI issue, it's a critical game engine issue and must be fixed.

It does not. The permanent keeps the SpellAbility, but it can only be activated from the hand (or other zones if something lets it be cast from there), same as other SpellAbilities.
